include/deal.II/matrix_free/evaluation_kernels_hanging_nodes.h

```diff
@@ -398,6 +421,1235 @@ namespace internal
+      }
+
+      static inline DEAL_II_ALWAYS_INLINE
+        std::array<MatrixFreeFunctions::ConstraintKinds, Number::size()>
+        create_mask(const std::array<MatrixFreeFunctions::ConstraintKinds,
+                                     Number::size()> mask)
+      {
+        return mask;
+      }
+
+      static inline DEAL_II_ALWAYS_INLINE Number
+      get_value(const Number &value, const index_type &)
+      {
+        return value;
+      }
+
+      static inline DEAL_II_ALWAYS_INLINE void
+      set_value(Number &result, const Number &value, const index_type &i)
+      {
+        (void)i;
+        result = value;
+      }
+    };
+
+    template <unsigned int side, bool transpose>
+    static inline DEAL_II_ALWAYS_INLINE void
+    interpolate_2D(
+      const unsigned int                                          given_degree,
+      const typename Trait<Number, VectorizationType>::index_type v,
+      const typename Trait<Number, VectorizationType>::interpolation_type
+        *DEAL_II_RESTRICT      weight,
+      Number *DEAL_II_RESTRICT values)
+    {
+      typename Trait<Number, VectorizationType>::value_type
+        temp[fe_degree != -1 ? (fe_degree + 1) : max_n_points_1D];
+
+      const unsigned int points =
+        (fe_degree != -1 ? fe_degree : given_degree) + 1;
+
+      AssertIndexRange(given_degree, max_n_points_1D);
+
+      const unsigned int d = side / 2; // direction
+      const unsigned int s = side % 2; // left or right surface
+
+      const unsigned int offset = dealii::Utilities::pow(points, d + 1);
+      const unsigned int stride =
+        (s == 0 ? 0 : (points - 1)) * dealii::Utilities::pow(points, d);
+
+      const unsigned int r1 = dealii::Utilities::pow(points, dim - d - 1);
+      const unsigned int r2 = dealii::Utilities::pow(points, d);
+
+      // copy result back
+      for (unsigned int i = 0, k = 0; i < r1; ++i)
+        for (unsigned int j = 0; j < r2; ++j, ++k)
+          temp[k] = Trait<Number, VectorizationType>::get_value(
+            values[i * offset + stride + j], v);
+
+      // perform interpolation point by point (note: r1 * r2 ==
+      // points^(dim-1))
+      for (unsigned int i = 0, k = 0; i < r1; ++i)
+        for (unsigned int j = 0; j < r2; ++j, ++k)
+          {
+            typename Trait<Number, VectorizationType>::value_type sum = 0.0;
+            for (unsigned int h = 0; h < points; ++h)
+              sum += Trait<Number, VectorizationType>::get_value(
+                       weight[(transpose ? 1 : points) * k +
+                              (transpose ? points : 1) * h],
+                       v) *
+                     temp[h];
+            Trait<Number, VectorizationType>::set_value(
+              values[i * offset + stride + j], sum, v);
+          }
+    }
+
+    template <unsigned int direction,
+              unsigned int d,
+              bool         transpose,
+              bool         skip_borders>
+    static inline DEAL_II_ALWAYS_INLINE void
```

Comment:
How big is the effect of inlining this function? I am hesitant to inline this in `debug` mode, but it might make sense in `release` mode. Can you wrap this into an `ifndef DEBUG`?